spark/src/test/scala/org/apache/spark/sql/delta/DeltaInsertIntoImplicitCastSuite.scala

```diff
@@ -0,0 +1,208 @@
+        DFv1Save(SaveMode.Append),
+        DFv1Save(SaveMode.Overwrite),
+        DFv2Append,
+        DFv2Overwrite,
+        DFv2OverwritePartition
+      ),
+      confs = Seq(DeltaSQLConf.DELTA_SCHEMA_AUTO_MIGRATE.key -> schemaEvolution.toString)
+    )
+
+    testInserts("insert with implicit up and down cast on fields nested in array, " +
+      s"schemaEvolution=$schemaEvolution")(
+      initialSchemaDDL = "key int, a array<struct<x: long, y: int>>",
+      initialJsonData = Seq("""{ "key": 1, "a": [ { "x": 1, "y": 2 } ] }"""),
+      partitionBy = Seq("key"),
+      overwriteWhere = "key" -> 1,
+      insertSchemaDDL = "key int, a array<struct<x: int, y: long>>",
+      insertJsonData = Seq("""{ "key": 1, "a": [ { "x": 3, "y": 4 } ] }"""),
+      expectedResult = ExpectedResult.Success(
+        expectedSchema = new StructType()
+          .add("key", IntegerType)
+          .add("a", ArrayType(new StructType()
+            .add("x", LongType)
+            .add("y", IntegerType, nullable = true)))),
+      // The following insert operations don't implicitly cast the data but fail instead - see
+      // following test covering failure for these cases. We should change this to offer consistent
+      // behavior across all inserts.
+      excludeInserts = Seq(
+        DFv1SaveAsTable(SaveMode.Append),
+        DFv1SaveAsTable(SaveMode.Overwrite),
+        DFv1Save(SaveMode.Append),
+        DFv1Save(SaveMode.Overwrite),
+        DFv2Append,
+        DFv2Overwrite,
+        DFv2OverwritePartition
+      ),
+      confs = Seq(DeltaSQLConf.DELTA_SCHEMA_AUTO_MIGRATE.key -> schemaEvolution.toString)
+    )
+
+    testInserts("insert with implicit up and down cast on fields nested in array, " +
+      s"schemaEvolution=$schemaEvolution")(
+      initialSchemaDDL = "key int, a array<struct<x: long, y: int>>",
+      initialJsonData = Seq("""{ "key": 1, "a": [ { "x": 1, "y": 2 } ] }"""),
+      partitionBy = Seq("key"),
+      overwriteWhere = "key" -> 1,
+      insertSchemaDDL = "key int, a array<struct<x: int, y: long>>",
+      insertJsonData = Seq("""{ "key": 1, "a": [ { "x": 3, "y": 4 } ] }"""),
+      expectedResult = ExpectedResult.Failure(ex => {
+        checkError(
+          ex,
+          errorClass = "DELTA_FAILED_TO_MERGE_FIELDS",
+          parameters = Map(
+            "currentField" -> "a",
+            "updateField" -> "a"
+          ))
+      }),
+      includeInserts = Seq(
+        DFv1SaveAsTable(SaveMode.Append),
+        DFv1SaveAsTable(SaveMode.Overwrite),
+        DFv1Save(SaveMode.Append),
+        DFv1Save(SaveMode.Overwrite),
+        DFv2Append,
+        DFv2Overwrite,
+        DFv2OverwritePartition
+      ),
+      confs = Seq(DeltaSQLConf.DELTA_SCHEMA_AUTO_MIGRATE.key -> schemaEvolution.toString)
+    )
+
+    testInserts("insert with implicit up and down cast on fields nested in map, " +
+      s"schemaEvolution=$schemaEvolution")(
+      initialSchemaDDL = "key int, m map<string, struct<x: long, y: int>>",
```

Comment:
Should we test the `map`'s key as well?

Reply:
This is going to be hard to test here, these tests rely on parsing json to setup the table and ingest data. and I don't believe that supports using anything other than a string for map keys.
That would be a pretty significant rewrite of all the tests